### ISSUE TITLE
Creating embedding is not reproducible

### ISSUE THREAD

**Issue by ericbroda** [stellargraph/stellargraph#508]
I am using StellarGraph to create embeddings for a particular graph/feature set.  Unfortunately, the embeddings are different each time I create/train the graph despite providing identical information each time.

Is this bug, or am I using StellarGraph incorrectly?

Below is the code that demonstrates the issue:

    import networkx as nx
    import random
    import numpy as np
    import pandas as pd
    import keras
    import stellargraph as sg
    from stellargraph.mapper import GraphSAGELinkGenerator, GraphSAGENodeGenerator
    from stellargraph.layer import GraphSAGE, link_classification
    from stellargraph.data import UnsupervisedSampler
    
    # Establish random seed
    RANDOM_SEED = 42
    random.seed(RANDOM_SEED)
    
    # Create a graph from well-known karate club data
    print(f"Creating graph")
    graph = nx.karate_club_graph()
    
    # Create features for each node
    print(f"Creating features")
    features = []
    nodes = list(graph.nodes)
    columns = ["c-" + str(x) for x in range(10)]
    nodes.sort()
    for node in nodes:
        f = {c: random.random() for c in columns}
        features.append(f)

    features_df = pd.DataFrame(features)
    print(f"features_df: \n{features_df}")
    
    for i in range(2):
        print(f"----- Iteration: {i} -----")
    
        # Create the model and generators
        print(f"Creating the model and generators")
        Gs = sg.StellarGraph(graph, node_features=features_df)
        unsupervisedSamples = UnsupervisedSampler(Gs, nodes=graph.nodes(), length=5, number_of_walks=3, seed=RANDOM_SEED)
        train_gen = GraphSAGELinkGenerator(Gs, 50, [5, 5], seed=RANDOM_SEED).flow(unsupervisedSamples)
        graphsage = GraphSAGE(layer_sizes=[100, 100], generator=train_gen, bias=True, dropout=0.0, normalize="l2")
        x_inp_src, x_out_src = graphsage.node_model(flatten_output=False)
        x_inp_dst, x_out_dst = graphsage.node_model(flatten_output=False)
    
        x_inp = [x for ab in zip(x_inp_src, x_inp_dst) for x in ab]
        x_out = [x_out_src, x_out_dst]
        edge_embedding_method = "l2"
        prediction = link_classification(output_dim=1, output_act="sigmoid", edge_embedding_method=edge_embedding_method)(x_out)
    
        # Create and train the Keras model
        model = keras.Model(inputs=x_inp, outputs=prediction)
        learning_rate = 1e-2
        model.compile(
            optimizer=keras.optimizers.Adam(lr=learning_rate),
            loss=keras.losses.binary_crossentropy,
            metrics=[keras.metrics.binary_accuracy])
    
        _ = model.fit_generator(train_gen, epochs=5, verbose=2, use_multiprocessing=False, workers=1, shuffle=False)
    
        # Create the embeddings
        print(f"Creating the embeddings")
        nodes = list(graph.nodes)
        nodes.sort()
        print(f"Nodes: {nodes}")
    
        # Create a generator that serves up nodes for use in embedding prediction / creation
        node_gen = GraphSAGENodeGenerator(Gs, 50, [5, 5], seed=RANDOM_SEED).flow(nodes)
    
        embedding_model = keras.Model(inputs=x_inp_src, outputs=x_out_src)
        embeddings = embedding_model.predict_generator(node_gen, workers=4, verbose=1)
        embeddings = embeddings[:, 0, :]
    
        np.set_printoptions(threshold=10)
        print(f"embeddings: {embeddings.shape} \n{embeddings}")

There are a number of debug (print output) statements when the code is executed. (sample output is shown below).  Note that the embeddings are different despite the identical inputs, graph configuration, model configuration, and random see values.

    ----- Iteration: 0 -----
    :
    :
    1/1 [==============================] - 0s 58ms/step
    embeddings: (34, 100) 
    [[-0.10566715  0.02253576 -0.18743701 ... -0.1028127   0.03689012
      -0.02482301]
     [-0.03171733  0.01606975 -0.08616363 ... -0.11775644  0.0429472
      -0.02371055]
     [-0.05802531  0.03910012 -0.10229243 ... -0.15050544  0.06637941
      -0.01950052]
     ...
     [ 0.03011296  0.08852117 -0.01836969 ... -0.154132    0.03844732
      -0.08643046]
     [ 0.01052345 -0.0123206   0.08913474 ... -0.11741614  0.03202919
      -0.04432516]
     [ 0.01951274  0.06263477  0.07959272 ... -0.10350229  0.05735112
      -0.0368157 ]]
    :
    :
    ----- Iteration: 1 -----
    embeddings: (34, 100) 
    [[ 0.11182436 -0.02642134  0.01168384 ...  0.10322241 -0.01680471
      -0.03918815]
     [ 0.02391489  0.02674667 -0.00091334 ...  0.12946768 -0.02389602
      -0.01414653]
     [ 0.08718258 -0.01711811 -0.05704292 ...  0.13477756 -0.00658288
      -0.05889895]
     ...
     [ 0.06843725 -0.13134597 -0.10870655 ...  0.11091235 -0.05146989
      -0.06138216]
     [-0.00593233 -0.05901312 -0.02113489 ... -0.01590953 -0.02516254
      -0.02280537]
     [ 0.00871993 -0.04059998 -0.07237951 ... -0.01590569 -0.00954109
      -0.01116194]]

**Comment by youph** [comment]
@annitrolla @adocherty @PantelisElinas can someone have a look please?
We’ve looked at this issue before, did we narrow it down to the optimiser? I remember that, with the seeds fixed, the random walks and the target-context pairs were the same, initial model weights were the same as well, but then the two models started to diverge while training (even after 1 epoch), for a reason we couldn’t find (everything we looked at was same between the models: initial state, training data, etc.). BUT when we changed the optimiser to the simple `sgd` instead of `adam`, the divergence of the two models vanished, and they became identical after training. Every other optimiser we tried (`nesterov`, `adam`, etc.) introduced the divergence between the models. Very strange. Seems to be some deeply rooted issue in keras?

**Comment by youph** [comment]
@ericbroda actually, your two models are not created equal: their initial weights are different. Before doing anything else, you may want to fix the random seed at the start of EACH iteration in your loop; see https://machinelearningmastery.com/reproducible-results-neural-networks-keras/

**Comment by adocherty** [comment]
> @ericbroda actually, your two models are not created equal: their initial weights are different. Before doing anything else, you may want to fix the random seed at the start of EACH iteration in your loop; see https://machinelearningmastery.com/reproducible-results-neural-networks-keras/
> 
> Also, I'd suggest upgrading your `stellargraph` library to the most recent version in the `develop` branch on github: https://github.com/stellargraph/stellargraph

Not that fixing the random seed in the `random` library does not fix the random seed for deep learning models. You also need fix the random seeds in tensorflow and numpy, as well as in the generators (as you have done).  See the articles above for more information on this.  However, it is difficult to control for all sources of randomness in deep learning models, as even machine precision noise can affect the optimizer.

Is there a particular reason why you need reproducible training?

**Comment by ericbroda** [comment]
@adocherty Reason for requiring reproducible training is to support reproducible clustering to support incremental topic modelling on large corpus that grows daily.  More specifically I am using StellarGraph to create a model then get embeddings which are then used for clustering to support topic modelling.  The data (corpus) grows daily and hence I use induction to continuously grow the model.  Unfortunately, each time I get the embeddings to perform cluster analysis on the new (inducted) data, the embeddings are very different and do not align at all properly.  As a test case, I tried to add (induct) *identical* data into an existing model and the embedding come out quite differently.  The net-result, is that without this being addressed, a fundamental (identified in the demos also) use-case that use the embeddings for downstream work for production-scale workload (ie. requires continuous updating of models) will not be addressed.

**Comment by ericbroda** [comment]
@youph I tried the SGD optimizer and it did not address the embeddings issue (ie. embeddings are still different).  As a side note, changing the optimizer gave measurable poorer accuracy in the model.

**Comment by ericbroda** [comment]
@youph I just tried setting the TensorFlow random seed at the beginning of the demo code (code below, as recommended in provided MachineLearningMastery link) and the embeddings still came out different.

```
RANDOM_SEED = 42
random.seed(RANDOM_SEED)
from tensorflow import set_random_seed
set_random_seed(RANDOM_SEED)
```

**Comment by ericbroda** [comment]
In the interest of completeness, I tried the following steps to ensure *all* random seeds across python/tensorflow/keras are set (see code below, put at the top of the demo code).  The embeddings still came out differently (ie. this did not solve the problem).

```
# Establish random seed
RANDOM_SEED = 42

# 1. Set PYTHONHASHSEED environment variable at a fixed value
import os
os.environ['PYTHONHASHSEED'] = str(RANDOM_SEED)

# 2. Set python built-in pseudo-random generator at a fixed value
import random
random.seed(RANDOM_SEED)

# 3. Set numpy pseudo-random generator at a fixed value
np.random.seed(RANDOM_SEED)

# 4. Set tensorflow pseudo-random generator at a fixed value
import tensorflow as tf
tf.set_random_seed(RANDOM_SEED)

# 5. Configure a new global tensorflow session
import keras
from keras import backend as K
session_conf = tf.ConfigProto(intra_op_parallelism_threads=1, inter_op_parallelism_threads=1)
sess = tf.Session(graph=tf.get_default_graph(), config=session_conf)
K.set_session(sess)
```

**Comment by ericbroda** [comment]
One more note:  I tried upgrading to StellarGraph version 0.8.0 and the demo code breaks (Note: the demo code (above, start of this issue) still conforms to pattern in StellarGraph clustering demo).

Any help or ideas are appreciated!

Code:
```
x_inp = [x for ab in zip(x_inp_src, x_inp_dst) for x in ab]
x_out = [x_out_src, x_out_dst]
prediction = link_classification(output_dim=1, output_act="sigmoid", edge_embedding_method="ip")(x_out)

# Create and train the Keras model
#### Breaks on the line below
model = keras.Model(inputs=x_inp, outputs=prediction)
```

With the following traceback:
```
Traceback (most recent call last):
  File "stellargraph_demo.py", line ..., in <module>
    model = keras.Model(inputs=x_inp, outputs=prediction)
  File "/Users/ericbroda/.virtualenvs/gnn/lib/python3.6/site-packages/keras/legacy/interfaces.py", line 91, in wrapper
    return func(*args, **kwargs)
  File "/Users/ericbroda/.virtualenvs/gnn/lib/python3.6/site-packages/keras/engine/network.py", line 93, in __init__
    self._init_graph_network(*args, **kwargs)
  File "/Users/ericbroda/.virtualenvs/gnn/lib/python3.6/site-packages/keras/engine/network.py", line 231, in _init_graph_network
    self.inputs, self.outputs)
  File "/Users/ericbroda/.virtualenvs/gnn/lib/python3.6/site-packages/keras/engine/network.py", line 1366, in _map_graph_network
    tensor_index=tensor_index)
  File "/Users/ericbroda/.virtualenvs/gnn/lib/python3.6/site-packages/keras/engine/network.py", line 1347, in build_map
    for i in range(len(node.inbound_layers)):
TypeError: object of type 'Activation' has no len()
```

**Comment by youph** [comment]
Re fixing all the seeds: even if all the seeds are fixed at the start of the script, the 2 models in the loop won’t be the same from the start. The easiest thing would be to build and train one model, run the script twice, and compare.

**Comment by ericbroda** [comment]
@youph Yes, that works (ie. the embeddings are consistent).

This may be a dumb question, but if the *all* the seeds (including TensorFlow seed) why is the model different each time?  Is this a TensorFlow/Keras issue?

**Comment by youph** [comment]
Fixing the seeds at the start of the script initialises the random generators to fixed states. Building and training the 1st model in the loop uses the random generators, so their internal states are updated, hence the 2nd model (that is built after the 1st without resetting all the random states) is necessarily different. Or are you fixing all the random states at the start of each loop iteration as well? 

(I’m on leave without a computer so can’t run the code)

**Comment by youph** [comment]
To achieve your ultimate goal of incremental representation learning on the updated corpus, you would need to save model weights between the updates, and restore them to that saved state before doing further sgd on updated data. Does this make sense?

**Comment by ericbroda** [comment]
Yes, this makes sense.  Thank you for all of your support!  I will be using your suggestions

**Comment by youph** [comment]
Let us know how it works out

**Comment by psyche11** [comment]
@ericbroda I have encountered the same problem. I have tried to save the model weights and to set all the random seeds, but I still couldn't get the same embeddings. Could you please tell me your solution in detail?

**Comment by kjun9** [comment]
@psyche11 @ericbroda This should now be resolved with the v0.9.0 release - unsupervised GraphSAGE can be used to create embeddings that are reproducible by fixing the seeds correctly! You can take a look at how this is being tested [here](https://github.com/stellargraph/stellargraph/blob/v0.9.0/tests/reproducibility/test_unsupervised_graphsage.py) to get an idea of how to create your own reproducible pipeline, but it will generally involve doing:
* fixing the seed for these external packages: `numpy`, `tensorflow`, and `random`
* providing a seed when constructing `UnsupervisedSampler` and `GraphSAGELinkGenerator` objects. These classes are used to perform random walks and neighbourhood sampling.

There's currently an issue (#684 ) to make it easier to make reproducible pipelines, and also one (#749 ) to test other algorithms for reproducibility, so hopefully this covers most use-cases.
